src/index.ts

```diff
@@ -1,17 +1,21 @@
-import shallowEqualArrays, { validArrayValue } from "./arrays";
-import shallowEqualObjects, { validObjectValue } from "./objects";
+import shallowEqualArrays from "./arrays";
+import shallowEqualObjects from "./objects";
 
-type Comparable<T> = Record<string, T> | T[] | null | undefined;
+type Comparable = Record<string, any> | any[] | null | undefined;
 
-function shallowEqual<T extends Comparable<T>>(a: T, b: T): boolean {
-  if (Array.isArray(a) || Array.isArray(b)) {
-    return shallowEqualArrays(a as validArrayValue<T>, b as validArrayValue<T>);
+function shallowEqual<T extends Comparable>(a: T, b: T): boolean {
+  const aIsArr = Array.isArray(a);
+  const bIsArr = Array.isArray(b);
+
+  if (aIsArr !== bIsArr) {
+    return false;
+  }
+
+  if (aIsArr) {
+    return shallowEqualArrays(a as any[], b as any[]);
   }
```

Comment:
LMK if this implementation feels cleaner / neater.

Reply:
(and if it feels ok to keep the `as` in this case, unless we find better ways to Typescript-differentiate objects from arrays)

Reply:
I can see that if we bump TypeScript to 4.4.4, it can actually narrow the type of `a` after checking `Array.isArray(a)`. So, we should be able to get rid of `as` if we upgrade TypeScript (feel free to make the call).
```typescript
const aIsArr = Array.isArray(a);
const bIsArr = Array.isArray(b);

if (aIsArr !== bIsArr) {
  return false;
}

if (aIsArr && bIsArr) {
  return shallowEqualArrays(a, b);
}

return shallowEqualObjects(a, b);
```

Reply:
Ah, ok -- very helpful. I've updated to Typescript 4.4.4, which required bumping Jest and TS-Jest (now both v29+). No type casting is needed, and the tests pass cleanly.

BTW, I find it interesting that Typescript requires `if (aIsArr && bIsArr)`, instead of just `if (aIsArray)`, given the logic in the if statement above it. But alas. So I've corrected that `if` to now also check both booleans.